File: server/scripts/fit-ovrweights.py
"""
Fit the game's actual overall formula per (position bucket, archetype) from the
game-generated classes dumped by dump-game-prospects.js, for the archetypes where
ovrweights.json disagrees with the game (offensive line, long snappers).

Model: OVR = 99 * (sum(w_i * attr_i)/sum(w) - DL) / (DH - DL), i.e. linear in the
attributes. We regress OVR on the entry's weighted attributes (non-negative least
squares), then express the fit back as weights + DesiredLow/DesiredHigh so the
existing calculator can apply it. Only archetypes whose fit beats the stored
entry by a clear margin are written to data/lookups/ovrweights-overrides[-m27].json.

  python scripts/fit-ovrweights.py [m26|m27]
"""
import json, sys, os
import numpy as np

def nnls(A, b, iters=3000):
    """Tiny projected-gradient non-negative least squares (no scipy needed)."""
    A = np.asarray(A, float); b = np.asarray(b, float)
    x = np.zeros(A.shape[1]); L = np.linalg.norm(A, 2) ** 2 or 1.0
    for _ in range(iters):
        g = A.T @ (A @ x - b)
        x = np.maximum(0.0, x - g / L)
    return x, None

version = (sys.argv[1] if len(sys.argv) > 1 else 'm26').lower()
here = os.path.dirname(os.path.abspath(__file__))
rows = json.load(open(os.path.join(here, '..', 'cache', f'game-prospects-{version}.json'), encoding='utf-8'))
ovrw = json.load(open(os.path.join(here, '..', 'data', 'lookups', 'ovrweights.json'), encoding='utf-8'))

POS_TO_BUCKET = {'QB':'QB','HB':'HB','FB':'FB','WR':'WR','TE':'TE','LT':'OT','RT':'OT','LG':'G','RG':'G','C':'C','LEDG':'DE','REDG':'DE','DT':'DT','SAM':'OLB','WILL':'OLB','MIKE':'MLB','CB':'CB','FS':'S','SS':'S','K':'KP','P':'KP','LS':'LS'}
STEM_OVERRIDE = {'BCVision':'ballCarrierVision','Press':'pressCoverage'}
def stem_to_attr(stem):
    return STEM_OVERRIDE.get(stem, stem[0].lower() + stem[1:])
def norm(s): return ''.join(ch for ch in s.lower() if ch.isalnum())

entries = {}
for e in ovrw:
    if not e.get('Pos'): continue
    w = {stem_to_attr(k[:-6]): float(v) for k, v in e.items() if k.endswith('Rating') and float(v or 0) != 0}
    suffix = norm(e['Archetype'].split('_', 1)[1] if '_' in e['Archetype'] else e['Archetype'])
    entries[(e['Pos'], suffix)] = {'weights': w, 'DL': float(e['DesiredLow']), 'DH': float(e['DesiredHigh'])}

def stored_ovr(e, r):
    s = sum(w * r.get(a, 0) for a, w in e['weights'].items()); sw = sum(e['weights'].values())
    return round(99 * (s / sw - e['DL']) / (e['DH'] - e['DL']))

groups = {}
for r in rows:
    bucket = POS_TO_BUCKET[r['pos']]
    suffix = norm(r['archetypeName'])
    key = None
    for (p, s) in entries:
        if p == bucket and (s == suffix or (suffix and (suffix in s or s in suffix))):
            key = (p, s); break
    if key is None:
        key = next(((p, s) for (p, s) in entries if p == bucket), None)
    if key is None: continue
    groups.setdefault(key, []).append(r)

overrides = {}
total = exact = 0
for key, rs in sorted(groups.items()):
    e = entries[key]
    n = len(rs)
    miss = sum(1 for r in rs if stored_ovr(e, r) != r['overall'])
    total += n; exact += n - miss
    if n < 12 or miss / n < 0.25: continue
    ALL = [k for k in rs[0].keys() if k not in ('pos','posId','archetype','archetypeName','overall')]
    def fit(attrs):
        Xa = np.array([[r.get(a, 0) for a in attrs] for r in rs], dtype=float)
        ya = np.array([r['overall'] for r in rs], dtype=float)
        xm = Xa.mean(axis=0); ym = ya.mean()
        c, _ = nnls(Xa - xm, ya - ym)
        c0 = ym - xm @ c
        pred = np.rint(Xa @ c + c0)
        return c, c0, int((pred == ya).sum()), int((np.abs(pred - ya) <= 1).sum())
    attrs = list(e['weights'].keys())
    c, c0, fit_exact, within1 = fit(attrs)
    # If the entry's own attributes cannot explain the game's overall (long snappers),
    # search all attributes and keep the ones the fit actually uses.
    # (Needs a big sample: 54 unknowns on 28 long snappers just fits noise.)
    if within1 < 0.7 * n and n >= 150:
        c2, c02, ex2, w12 = fit(ALL)
        if w12 > within1:
            keep = [a for a, ci in zip(ALL, c2) if ci > 0.02]
            c, c0, fit_exact, within1 = fit(keep)
            attrs = keep
    # Express as weights summing to 10 and DL/DH: OVR = 99*(wavg - DL)/(DH-DL), wavg = sum(w_i a_i)/10
    s = c.sum()
    if s <= 0: continue
    w = {a: round(10 * ci / s, 3) for a, ci in zip(attrs, c) if ci > 1e-9}
    # wavg = (c @ a) * 10 / s / 10 = (c @ a)/s ; OVR = 99*(wavg - DL)/(DH-DL) => 99/(DH-DL) = s ; -99*DL/(DH-DL) = c0
    span = 99 / s; DL = -c0 / s; DH = DL + span
    improved = fit_exact > (n - miss) + max(2, n // 10)
    print(f"{key[0]:3} {key[1]:14} n={n:3} stored exact {n-miss:3}  fit exact {fit_exact:3} (within1 {within1:3})  DL/DH {e['DL']:.0f}/{e['DH']:.0f} -> {DL:.2f}/{DH:.2f} {'WRITE' if improved else ''}")
    if improved:
        overrides[f"{key[0]}:{key[1]}"] = {'pos': key[0], 'archetype': key[1], 'weights': w, 'desiredLow': round(DL, 3), 'desiredHigh': round(DH, 3), 'n': n, 'fitExact': fit_exact, 'storedExact': n - miss}

print(f"{version}: stored formula exact {exact}/{total}")
dest = os.path.join(here, '..', 'data', 'lookups', 'ovrweights-overrides-m27.json' if version == 'm27' else 'ovrweights-overrides.json')
json.dump({'_source': f'fit from {total} game-generated {version} prospects (scripts/fit-ovrweights.py)', 'overrides': overrides}, open(dest, 'w', encoding='utf-8', newline='\n'), indent=1)
print('wrote', dest, len(overrides), 'overrides')

File: server/scripts/fit-ovrweights.js
/**
 * Check our OVR recompute (OVRWeightsCalculator) against the overall the game
 * itself wrote in its generated classes, per (position, archetype). Where the
 * formula is systematically off, fit DesiredLow/DesiredHigh from the data
 * (OVR = 99 * (weightedAvg - DL) / (DH - DL) is linear in weightedAvg) and write
 * data/lookups/ovrweights-overrides.json, which OVRWeightsCalculator applies.
 *
 *   node scripts/fit-ovrweights.js [m26|m27]
 */
require('tsx/cjs');
const fs = require('fs');
const os = require('os');
const path = require('path');
const { MdcService } = require('../src/services/MdcService');
const { Mdc27Service } = require('../src/services/Mdc27Service');
const { OVRWeightsCalculator } = require('../src/services/OVRWeightsCalculator');
const { LookupService } = require('../src/services/LookupService');
const { PositionMapper } = require('../src/services/PositionMapper');

const version = (process.argv[2] || 'm26').toLowerCase();
const home = process.env.USERPROFILE || os.homedir();
const dir = version === 'm27' ? `${home}/Documents/Madden NFL 27/saves` : `${home}/Documents/Madden NFL 26/Saves`;
const names = version === 'm27' ? ['TEST1', 'TEST2', 'TEST3', 'TESTSUPERSTRONG'] : ['RANDOMGEN1', 'RANDOMGEN2', 'RANDOMGEN3', 'RANDOMGEN4', 'RANDOMGEN5'];
const silence = (fn) => { const c = { ...console }; console.log = console.error = console.warn = console.info = () => {}; try { return fn(); } finally { Object.assign(console, c); } };

const groups = new Map(); // key pos:arch -> { xs: weightedAvg[], ys: ovr[] }
let total = 0, exact = 0, within1 = 0;
for (const n of names) {
  const f = path.join(dir, `CAREERDRAFT-${n}`);
  if (!fs.existsSync(f)) continue;
  const ps = silence(() => (version === 'm27' ? Mdc27Service.parse : MdcService.parse).call(version === 'm27' ? Mdc27Service : MdcService, fs.readFileSync(f))).filter((p) => p.firstName);
  for (const p of ps) {
    const posId = Number(p.position), arch = Number(p.archetype) || 0, ovr = Number(p.overall);
    const entry = OVRWeightsCalculator.ovrEntryFor(posId, arch);
    if (!entry || !entry.sumWeight) continue;
    let sum = 0;
    for (const [a, w] of Object.entries(entry.weights)) sum += (Number(p[a]) || 0) * w;
    const wavg = sum / entry.sumWeight;
    const got = OVRWeightsCalculator.computeOverall(posId, arch, p);
    total++; if (got === ovr) exact++; if (Math.abs(got - ovr) <= 1) within1++;
    const key = `${posId}:${arch}`;
    const g = groups.get(key) ?? { posId, arch, xs: [], ys: [], miss: 0, n: 0, entry };
    g.xs.push(wavg); g.ys.push(ovr); g.n++; if (got !== ovr) g.miss++;
    groups.set(key, g);
  }
}
console.log(`${version}: ${total} prospects, exact ${exact} (${((exact / total) * 100).toFixed(1)}%), within 1: ${within1}`);

const overrides = {};
for (const g of groups.values()) {
  if (g.n < 8 || g.miss / g.n < 0.25) continue;
  // least squares OVR = a*wavg + b
  const n = g.n, mx = g.xs.reduce((s, x) => s + x, 0) / n, my = g.ys.reduce((s, y) => s + y, 0) / n;
  let sxy = 0, sxx = 0;
  for (let i = 0; i < n; i++) { sxy += (g.xs[i] - mx) * (g.ys[i] - my); sxx += (g.xs[i] - mx) ** 2; }
  if (!sxx) continue;
  const a = sxy / sxx, b = my - a * mx;
  const DL = -b / a, DH = DL + 99 / a;
  // verify the fit
  let fitExact = 0;
  for (let i = 0; i < n; i++) if (Math.round(((g.xs[i] - DL) / (DH - DL)) * 99) === g.ys[i]) fitExact++;
  const posName = PositionMapper.name(g.posId), archName = LookupService.idToName('archetype', g.arch) || `#${g.arch}`;
  console.log(`  ${posName.padEnd(5)} ${archName.padEnd(22)} n=${String(g.n).padStart(3)} miss ${String(g.miss).padStart(3)} | stored DL/DH ${g.entry.desiredLow}/${g.entry.desiredHigh} -> fit ${DL.toFixed(2)}/${DH.toFixed(2)} (fit exact ${fitExact}/${n})`);
  if (fitExact > g.n - g.miss) overrides[`${g.posId}:${g.arch}`] = { pos: posName, archetype: archName, desiredLow: +DL.toFixed(3), desiredHigh: +DH.toFixed(3), n: g.n, fitExact };
}
const out = path.join(__dirname, '..', 'data', 'lookups', version === 'm27' ? 'ovrweights-overrides-m27.json' : 'ovrweights-overrides.json');
fs.writeFileSync(out, JSON.stringify({ _source: `fit from ${total} game-generated ${version} prospects`, overrides }, null, 1));
console.log(`wrote ${out} (${Object.keys(overrides).length} overrides)`);
